### PR TITLE
WFLY-6602 JGroups subsystem does not expose a proper capability contract

### PR DESCRIPTION
**Pull request by pferraro**
https://issues.jboss.org/browse/WFLY-6602

Replaces channel/stack ServiceName factories with corresponding capability/requirement ServiceName factories.

Also, adds a new DefaultableCapabilityReference implementation that can reference a DefaultableUnaryRequirement.  This allows an attribute to reference a dynamic requirement if specified, and a static requirement if unspecified.  Used to support Infinispan <transport channel="..."/>, where the channel references the default channel capability if left unspecified.

### CODE REVIEW

File: clustering/jgroups/extension/src/test/resources/org/jboss/as/clustering/jgroups/subsystem/subsystem-jgroups-transform-reject.xml [hunk 1]
```diff
@@ -75,12 +76,7 @@
             </relay>
         </stack>
         <stack name="minimal">
-            <transport type="UDP">
-                <default-thread-pool min-threads="11"/>
-                <internal-thread-pool queue-length="22"/>
-                <oob-thread-pool keepalive-time="34"/>
-                <timer-thread-pool max-threads="43"/>
```
Comment (bstansberry) [inline clustering/jgroups/extension/src/test/resources/org/jboss/as/clustering/jgroups/subsystem/subsystem-jgroups-transform-reject.xml]: This kind of thing always makes me nervous in a transformer test; i.e. that the thing being removed may be the thing that needed transformation.

A simple "It's not" reply will make me calm. ;)

**Comment by kabir** [comment]
@rachmatowicz @rhusar Can one of you please approve the changes?

**Comment by bstansberry** [comment]
It would be really great if these kinds of big changes can be broken up into several commits, with the 'refactoring' type stuff separate from the significant functional changes. It's nearly impossible for people to provide much of a meaningful review otherwise.

I know that can be hard to do sometimes, but it is possible and when I force myself to do it I usually am happier with the final result.

**Comment by pferraro** [comment]
@bstansberry Point taken.

File: clustering/jgroups/extension/src/test/resources/org/jboss/as/clustering/jgroups/subsystem/subsystem-jgroups-transform-reject.xml (hunk 1 above)
Comment (pferraro) [inline clustering/jgroups/extension/src/test/resources/org/jboss/as/clustering/jgroups/subsystem/subsystem-jgroups-transform-reject.xml]: WFLY-6766 makes validation of the rejection of these resources impossible.  Consequently, I need to either remove these items from the configuration being tested, or ignore the whole test.

Comment (rhusar) [inline clustering/jgroups/extension/src/test/resources/org/jboss/as/clustering/jgroups/subsystem/subsystem-jgroups-transform-reject.xml]: Well, the rejections are still being tested in the "maximal" stack (i.e. https://github.com/wildfly/wildfly/pull/8978/files/bc5870ef9e54db25613e3dbff253d682af777329#diff-4e1952a60eaff7ba9b44979ec09e94c4R43), so some rejection test are still there. Hm, I don't remember why I added these here anyway.

**Comment by kabir** [comment]
@pferraro Probably needs some capabilities set up in the new messaging subsystem test case

**Comment by pferraro** [comment]
@kabir Thanks for the heads up.  Fixing now.

**Comment by kabir** [comment]
Also needs the fixes to be backported before it can be merged

**Comment by pferraro** [comment]
@kabir Done.

**Comment by rhusar** [comment]
👍

**Pull request merged**